source/auth/auth.rst

```diff
@@ -1159,6 +1159,274 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+
+If the refresh callback is given and the request callback is not given,
+the driver MUST raise an error.  If DEVICE_NAME is given and one or more
+callbacks are given, the driver MUST raise an error.
+
+If no callbacks are given, the driver MUST enforce that a DEVICE_NAME
+mechanism_properties is set and one of ("aws", "azure", or "gcp").
+The callback mechanism can be used to support both Authentication
+Code Workflows or Device workflows that are not explicitly implemented
+by drivers.  If there is no callback and no DEVICE_NAME, or the
+DEVICE_NAME is set but credentials cannot be automatically obtained,
+the driver MUST raise an error.
+
+
+Supported Device Workflows
+``````````````````````````
+
+Drivers MUST support device workflows for "aws", "azure", and "gcp", given
+by the DEVICE_NAME mechanism property.  In all cases the acquired token
+will be given as the ``jwt`` argument and the second client step of the
+OIDC SASL exchange MUST be made directly, skipping the clientStep1.
+Drivers MUST raise an error if both a DEVICE_NAME and PRINCIPAL_NAME are
+given, since the device workflow will not use the PRINCIPAL_NAME.
+
+AWS
+___
+
+When the DEVICE_NAME mechanism property is set to "aws", the driver MUST
+attempt to read the value given by the ``AWS_WEB_IDENTITY_TOKEN_FILE`` and
+interpret it as a file path.  The contents of the file are read as the
+access token.
+
+Azure
+_____
+
+When the DEVICE_NAME mechanism property is set to "azure", the driver MUST
+acquire an access token using the workflow described in "Obtaining an Access Token for Azure Key Vault" <TODO insert link to client-side-encryption>.
+In the case of authorization failure, the cache used for azure credentials
+MUST be cleared.
+
+GCP
+___
+
+When the DEVICE_NAME mechanism property is set to "gcp", the driver MUST
+acquire an access token using the workflow described in "Obtaining GCP Credentials" <TODO insert link to client-side-encryption>.
+
+
+Caching Credentials
+```````````````````
+
+Drivers MUST enable caching when callback(s) are provided to the mongo client.
+When an authorization request is made and there is a valid cached response,
+the driver MUST use the cached authorization token if it has not expired.
+
+A cache value is considered valid if it has been created in the past 5 hours.
+The cache is kept alive to preserve the serverStep1 response, as well as
+account for refresh tokens, which typically have an (unknown) lifetime that
+is longer than the access token lifetime.  The refresh token may either be
+stored by the user application, or as part of the request/refresh token
+response in the driver cache. To prevent a memory leak, the driver MUST clear
+invalid cache values at a regular interval, or during every authentication
+attempt.
+
+The cache keys MUST include the principal name (or empty string) and the
+resolved host name for the current server.  Using the resolved host name
```

Comment:
This should clarify what “resolved host name” means specifically (or link to some definition if this is a common term in driver specs) … the text below makes it sounds like it’s the hostname of the primary for replica sets?

Reply:
Updated to "DNS resolved name".

Reply:
It’s honestly still not really clear what this means… Does this mean the server’s IP address? Is the cache for a single endpoint, or for an entire replicaset? Does “DNS resolved” mean CNAME resolution? Does it include port information? etc.